File: web/knight/color.js

function makeHslString(disp) {
	return "hsl(" + disp.hue + ", " + disp.saturation + "%, " + disp.lightness + "%)";
}

function updateDisplay() {
	this.hue = Math.max(0, Math.min(this.hue, 360));
	this.saturation = Math.max(0, Math.min(this.saturation, 100));
	this.lightness = Math.max(0, Math.min(this.lightness, 100));
	document.getElementById(this.id + "-hue").value = this.hue;
	document.getElementById(this.id + "-saturation").value = this.saturation;
	document.getElementById(this.id + "-lightness").value = this.lightness;
	document.getElementById(this.id + "-color").setAttribute("style", "background-color:" + makeHslString(this));
}

function createColorDisplay(id) {
	
	var h = document.getElementById(id + "-hue");
	var s = document.getElementById(id + "-saturation");
	var l = document.getElementById(id + "-lightness");
	if (!h || !s || !l) return;

	var disp = {
		id: id,
		hue: parseInt(h.value),
		saturation: parseInt(s.value),
		lightness: parseInt(l.value),
		update: updateDisplay
	}
	var container = document.getElementById(id + "-color-container");
	if (!container) return;

	var color = document.createElement("div");
	color.setAttribute("id", id + "-color");
	color.setAttribute("class", "colordisplay");
	container.appendChild(color);
	
	h.addEventListener("change", function (evt) {
		disp.hue = evt.srcElement.value;
		disp.update();
	});

	s.addEventListener("change", function (evt) {
		disp.saturation = evt.srcElement.value;
		disp.update();
	});
	l.addEventListener("change", function (evt) {
		disp.lightness = evt.srcElement.value;
		disp.update();
	});
	
	disp.update();
	
	return disp;	
}

File: web/server/color.go
package main

import (
	"errors"
)

type Color struct {
	v1, v2, v3 int
}

func MakeColor(v1, v2, v3 int) Color {
	return Color{v1: v1, v2: v2, v3: v3}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func HslToHsv(c Color) (Color, error) {
	if c.v1 < 0 || c.v1 > 360 {
		return c, errors.New("Hue out of range")
	}
	if c.v2 < 0 || c.v2 > 100 {
		return c, errors.New("Saturation out of range")
	}
	if c.v3 < 0 || c.v3 > 100 {
		return c, errors.New("Lightness out of range")
	}
	v := c.v3 + ((c.v2 * min(c.v3, 100 - c.v3))/100)
	s := 0
	if v != 0 {
		s = 200 - (200 * c.v3/v)
	}
	c.v2 = s
	c.v3 = v
	return c, nil
}

func (c *Color) ToOpenCVRange() {
	c.v1 = c.v1 / 2
	c.v2 = (c.v2 * 255) / 100
	c.v3 = (c.v3 * 255) / 100
}
